Draft: Feat/catch2 compat

CrustyAuklet opened the pull request:
Very rought draft, but I wanted to open it so others can look and comment.

This is work to make the Catch2 XML output match up with Catch2. Like the other open MRs I notices this when integrating with my teams IDE and (more importantly) CI test reporting systems. 

Without these changes CI and IDEs work ok, but failures in a section just fail the whole test. When we were using Catch2 a failure would be specific to a section allowing us to quickly drill down to the problem. So looking into the difference it is just the lack of section output in the "high" verbosity mode. Catch2 outputs section XML data by default.

It was fairly easy to make it work with basic testing, pass and fail. There are more complex edge cases around skips, nested sections, exceptions being thrown, and different verbosity levels. I am still ironing those out and need to make sure there is testing for it all.

As I have time this week i will add some examples here, and some pictures to show the different in CI/IDEs.

cschreib commented:
Thanks for spotting this. I could have sworn I got the output for sections to match Catch2 when I last tested this; but perhaps I had looked at a too simple case and missed the bigger picture. 

> As I have time this week i will add some examples here, and some pictures to show the different in CI/IDEs.

It would be very interesting to compare the actual XML output between the two in a real-world case; the test project I use for testing and benchmarks doesn't actually use sections.

cschreib commented on tests/approval_tests/data/expected/reporter_catch2_xml_allfail:
```diff
@@ -145,38 +145,41 @@
+      <OverallResults successes="0" failures="0" expectedFailures="0" skipped="true" durationInSeconds="*"/>
     </Section>
-    <OverallResult success="false" durationInSeconds="*"/>
+    <Failure filename="*testing_reporters.cpp" line="*">
+      unexpected std::exception caught; message: unexpected error
+    </Failure>
```
I think this message is printed too late; I believe we want it to appear inside the `<Section>` where the exception was thrown, as it was before. See comment in `~section_entry_checker()`.

cschreib replied:
Fixed.

cschreib commented on src/snitch_section.cpp:
```diff
@@ -15,12 +16,20 @@ section_entry_checker::~section_entry_checker() {
             // We are unwinding the stack because an exception has been thrown;
             // avoid touching the section state since we will want to report where
             // the exception was thrown.
+            state.reg.report_callback(
+                state.reg, event::section_ended{
+                               state.sections.current_section.back().id,
+                               state.sections.current_section.back().location, true});
```
See comment in the approval tests; I think this is done too soon. When an exception is in flight, it has not been reported yet, but we now close the section immediately. This results in all unhandled exceptions being reported at the root of the test case, rather than inside the section from which it originated. 

I believe this particular case (close the section when unwinding an exception) can be handled in the registry instead, to solve this problem. That would be here: https://github.com/snitch-org/snitch/blob/d6632123cc8d13bdbc5cd60fd6741b9e0f635e82/src/snitch_registry.cpp#L554-L558

cschreib replied:
Fixed.

cschreib commented on include/snitch/snitch_section.hpp:
```diff
@@ -3,12 +3,21 @@
 #include "snitch/snitch_config.hpp"
 #include "snitch/snitch_test_data.hpp"
+#if SNITCH_WITH_TIMINGS
+#    include <chrono>
```
We had, so far, managed to avoid including `<chrono>` in public headers. Could we keep it that way, to keep compilation time down? I just ran some tests, measuring GCC time to parse an empty `.cpp` with `#include`:
 - include `snitch_all.hpp` as of today's `main` branch: 38ms.
 - same, but add `<chrono>`: 46ms (+8ms, or +22%).

This feels like a high price to pay for just storing a time point. We could perhaps store the numerical value of the time point in the `struct` here, as an anonymous `std::size_t` or similar that is `>= sizeof(std::chrono::steady_clock::rep)`, and then convert back and forth to actual time points in the `*.cpp` files.

CrustyAuklet replied:
Makes sense, I see that <type_traits> is included already so I use that instead to create a signed version of `size_t` (since not all platforms define ssize_t). I added some aliases in the snitch::impl for chrono in the cpp file as well to make the chrono code less verbose and avoid bugs if anyone wants to change clocks or duration types in the future.

cschreib replied:
I ended up adding a new file `snitch_time.cpp` to wrap this up, since we needed chrono functionality elsewhere too (`registry`). In the end, all we need is a simple API with a function returning the "current time point" in some unspecified unit (I picked "number of nanoseconds since test run start" -- always positive with a `steady_clock` so a `std::size_t` is OK, and will only wrap around after ~600 years of execution, also OK) and another function to calculate the time difference between two time points.

cschreib commented on src/snitch_reporter_teamcity.cpp:
```diff
@@ -148,6 +148,8 @@ void report(const registry& r, const snitch::event::data& event) noexcept {
                 send_message(r, "testFinished", {{"name", make_full_name(e.id)}});
 #    endif
             },
+            [&](const snitch::event::section_started&) {},
+            [&](const snitch::event::section_ended&) {},
```
TeamCity uses the following:
 - `##teamcity[blockOpened name='...' description='...']`
 - `##teamcity[blockClosed name='...']`

CrustyAuklet replied:
In the  `print_assertion()` is see where it iterates over the sections, but I don't see how it outputs the test you show here. I don't see and "blockOpened" or "blockCosed"?

cschreib replied:
Got it :)

cschreib commented on src/snitch_section.cpp:
```diff
@@ -15,12 +16,20 @@ section_entry_checker::~section_entry_checker() {
 
         pop_location(state);
 
+        asserts          = state.asserts - asserts;
+        failures         = state.failures - failures;
+        allowed_failures = state.allowed_failures - allowed_failures;
```
This gives two different meanings to `section_entry_checker::asserts`/`failures`/`allowed_failures` depending on context: on creation they hold the initial state, and on destruction they hold the difference. This will be a source of confusion, I fear. But in fact, we only need to store the initial state (and should probably rename the member variables as such, e.g. `initial_sate.asserts`/`initial_state.failures`/...); the difference computed here could be stored in local variables.

CrustyAuklet replied:
> Alternatively, we could instead store the actual number of assertions (etc) that were recorded, and make the registry propagate the counts to all open sections in register_assertion(). This is a little bit more work for the CPU, but the code might be simpler.

This also might be necessary since I am noticing in my tests that Catch2 counts these things cumulatively for nested sections but with this strategy snitch does not.

given
```c++
SECTION("Section1") {
    CHECK(true);
    SECTION("Section1.1") {
        CHECK(false);
    }
}
```

this code outputs
```
    <Section name="Section1" filename="all_fail.cpp" line="10">
      <Section name="Section1.1" filename="all_fail.cpp" line="12">
        <Expression success="false" type="CHECK" filename="all_fail.cpp" line="13">
          <Original>
            false
          </Original>
          <Expanded>
            false
          </Expanded>
        </Expression>
        <OverallResults successes="0" failures="1" expectedFailures="0" skipped="false" durationInSeconds=<TIME>/>
      </Section>
      <OverallResults successes="0" failures="0" expectedFailures="0" skipped="false" durationInSeconds=<TIME>/>
    </Section>
```

Catch2:
```
    <Section name="Section1" filename="all_fail.cpp" line="10">
      <Section name="Section1.1" filename="all_fail.cpp" line="12">
        <Expression success="false" type="CHECK" filename="all_fail.cpp" line="13">
          <Original>
            false
          </Original>
          <Expanded>
            false
          </Expanded>
        </Expression>
        <OverallResults successes="0" failures="1" expectedFailures="0" skipped="false" durationInSeconds=<TIME>/>
      </Section>
      <OverallResults successes="1" failures="1" expectedFailures="0" skipped="false" durationInSeconds=<TIME>/>
    </Section>
```

CrustyAuklet replied:
if we are counting the actual count for each section, then that needs to happen in the registry too? I don't see a way to do it in the `section_entry_checker` since it is only called on entry and exit.

cschreib replied:
Yep; I've implemented this so the registry does the propagation of assert counts to all currently open sections.

cschreib commented on src/snitch_section.cpp:
```diff
@@ -89,6 +117,11 @@ section_entry_checker::operator bool() {
          state.sections.depth < state.sections.levels.size())) {
 
+        // First time entering this section, emit the section start event
+        if (level.current_section_id == level.previous_section_id + 1 &&
+            level.current_section_id != level.previous_section_id) {
```
I don't think this is right; the second part of the expression after the `&&` will always evaluate to `true` if the first part also evaluates to `true`. Perhaps this simply needs to be removed:
```suggestion
        if (level.current_section_id == level.previous_section_id + 1) {
```

cschreib commented on src/snitch_section.cpp:
```diff
@@ -29,6 +38,18 @@ section_entry_checker::~section_entry_checker() {
+#if SNITCH_WITH_TIMINGS
+            const auto  end_time = std::chrono::steady_clock::now();
+            const float duration = std::chrono::duration<float>(end_time - start_time).count();
+            state.reg.report_callback(
+                state.reg,
+                event::section_ended{
+                    data.id, data.location, false, asserts, failures, allowed_failures, duration});
+#else
+            state.reg.report_callback(
+                state.reg,
+                event::section_ended{data.id, data.location, asserts, failures, allowed_failures});
```
I think all the reporting logic should be outsourced to `registry`; the `report_callback` isn't really meant to be used outside the `registry` class (it is exposed as a public member primarily so it can be assigned, otherwise it would be private). We probably want helper members functions like `registry::report_section_entered` / `registry::report_section_exited` to achieve this.

CrustyAuklet replied:
This sounds much better. Adding the reporting here was the stuff I was most uncomfortable with and it seems very hacky.

cschreib replied:
Done

CrustyAuklet commented:
thanks for the feedback, I will look it over in depth this coming week.

I've been working on making tests and comparing Catch2  to Snitch output in a repo on my account here: https://github.com/CrustyAuklet/snitch-xml-output. So far it's pretty close but I will have to add some more variations, especially to cover the exception stuff.

cschreib commented:
Hi there! Just wondering if this was still on your mind. Let me know if not, and I can take over.

CrustyAuklet commented:
It is, I have just had a ton on my plate. I am going to try to re-familiarize myself with it this afternoon and see if have any questions on your feedback.

I won't be offended at all if you think it's an easy fix for you to do and you want to get it done 😄

CrustyAuklet commented:
fixed some of the small easy stuff and rebased, but spent a lot of time running some example tests in that other repo with exceptions. I think most of this code needs to be moved to the registry like you said.

cschreib commented:
Thanks for looking at it further. Does snitch reporting works as expected in your tests now? If so, I can do another pass of review, and if all looks fine, I can also take care of the final refactoring.

FYI: I just hit some of the issues you were having, so I can confirm (if this was even needed!) that things are currently broken on the main branch. I'm implementing a test runner for SublimeText, and for the sake of the exercise, I implemented a test adaptor for *Catch2* first, and then tried to use it with *snitch* without touching it. Didn't work out of the box! I'm hoping this will be solved with your PR.

CrustyAuklet commented:
Without exceptions it works as expected and I am using it on some of my work projects. But with exceptions It doesn't match up with the output of Catch2, and that is where I got a bit stuck and maybe the code does need to move into the registry as you suggested.

The test using exceptions in the repo https://github.com/CrustyAuklet/snitch-xml-output shows some of the differences. Catch2 is more verbose but the real issue is that is seems that exceptions thrown from within sections are not output until the parent section. 

Here are the output XML files from that test:
[exceptions.catch2.txt](https://github.com/user-attachments/files/17017420/exceptions.catch2.txt)
[exceptions.snitch.txt](https://github.com/user-attachments/files/17017421/exceptions.snitch.txt)

cschreib commented:
I believe the remaining issues are now sorted if you try the branch [cschreib/catch2-compat](https://github.com/snitch-org/snitch/tree/cschreib/catch2-compat). I created a PR for it https://github.com/snitch-org/snitch/pull/183 to get the CI runs. If you have no objection, I suggest switching to the new PR and closing this one (unless you feel like doing the git magic to pull my changes back into your branch; you'll need to rebase though). If you wanted to review the changes I pushed over your last commit, feel free!

cschreib commented:
Superseded by https://github.com/snitch-org/snitch/pull/183; thanks again for kick-starting this!
